TestReduceDataForward_Sum has data race

`TestReduceDataForward_Sum` is failing because it is publishing and reading concurrently. This behaviour only happens in the test because we run both producer and consumer at the same time. 

I prefer not to add locks in publisher.go because it will hurt the performance

```==================
WARNING: DATA RACE
Write at 0x00c0002de278 by goroutine 422:
  github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).validateWatermark()
      /home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:145 +0x49e
  github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).PublishWatermark()
      /home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:99 +0x92
  github.com/numaproj/numaflow/pkg/reduce.publishMessages()
      /home/runner/work/numaflow/numaflow/pkg/reduce/data_forward_test.go:873 +0x63b
  github.com/numaproj/numaflow/pkg/reduce.TestReduceDataForward_Sum.func6()
      /home/runner/work/numaflow/numaflow/pkg/reduce/data_forward_test.go:407 +0xef

Previous read at 0x00c0002de278 by goroutine 416:
  github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).validateWatermark()
      /home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:147 +0x55e
  github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).PublishIdleWatermark()
      /home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:159 +0xbb
  github.com/numaproj/numaflow/pkg/reduce/pnf.(*ProcessAndForward).publishWM()
      /home/runner/work/numaflow/numaflow/pkg/reduce/pnf/processandforward.go:269 +0x483
  github.com/numaproj/numaflow/pkg/reduce/pnf.(*ProcessAndForward).Forward()
      /home/runner/work/numaflow/numaflow/pkg/reduce/pnf/processandforward.go:150 +0x82c
  github.com/numaproj/numaflow/pkg/reduce/readloop.(*orderedForwarder).forward()
      /home/runner/work/numaflow/numaflow/pkg/reduce/readloop/ordered.go:183 +0x1473
  github.com/numaproj/numaflow/pkg/reduce/readloop.newOrderedForwarder.func1()
      /home/runner/work/numaflow/numaflow/pkg/reduce/readloop/ordered.go:71 +0x58

Goroutine 422 (running) created at:
  github.com/numaproj/numaflow/pkg/reduce.TestReduceDataForward_Sum()
      /home/runner/work/numaflow/numaflow/pkg/reduce/data_forward_test.go:407 +0x8f9
  testing.tRunner()
      /opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1493 +0x47

Goroutine 416 (running) created at:
  github.com/numaproj/numaflow/pkg/reduce/readloop.newOrderedForwarder()
      /home/runner/work/numaflow/numaflow/pkg/reduce/readloop/ordered.go:71 +0x378
  github.com/numaproj/numaflow/pkg/reduce/readloop.NewReadLoop()
      /home/runner/work/numaflow/numaflow/pkg/reduce/readloop/readloop.go:81 +0xc4
  github.com/numaproj/numaflow/pkg/reduce.NewDataForward()
      /home/runner/work/numaflow/numaflow/pkg/reduce/data_forward.go:79 +0x2d5
  github.com/numaproj/numaflow/pkg/reduce.TestReduceDataForward_Sum()
      /home/runner/work/numaflow/numaflow/pkg/reduce/data_forward_test.go:399 +0x669
  testing.tRunner()
      /opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1493 +0x47
==================
{"level":"info","ts":1677885385.6130238,"logger":"numaflow","caller":"readloop/readloop.go:408","msg":"Close of book","partitionID":"120000-240000"}
{"level":"warn","ts":1677885385.6133683,"logger":"numaflow","caller":"publish/publisher.go:147","msg":"Skip publishing the new watermark because it's older than the current watermark","entityID":"source-reduce-buffer","otStore":"TestReduceDataForward_Sum_OT","hbStore":"TestReduceDataForward_Sum_PROCESSORS","head":"1970-01-01T00:04:59Z","new":"1970-01-01T00:02:00Z"}
    testing.go:1319: race detected during execution of test
--- FAIL: TestReduceDataForward_Sum (0.15s)
```

--- a/pkg/reduce/data_forward_test.go
+++ b/pkg/reduce/data_forward_test.go
@@ -296,6 +296,7 @@ func TestReduceDataForward_Count(t *testing.T) {
 		startTime      = 60000 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -312,18 +313,19 @@ func TestReduceDataForward_Count(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publisherMap := createPublisherForBuffer(ctx, toBuffer, pipelineName)
 
 	// create a fixed window of 60s
 	window := fixed.NewFixed(60 * time.Second)
 
 	var reduceDataForward *DataForward
-	reduceDataForward, err = NewDataForward(ctx, CounterReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(ctx, CounterReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publisherMap,
 		window, WithReadBatchSize(10))
 	assert.NoError(t, err)
 
@@ -331,7 +333,7 @@ func TestReduceDataForward_Count(t *testing.T) {
 	go reduceDataForward.Start()
 
 	// start the producer
-	go publishMessages(ctx, startTime, messageValue, 300, 10, p[fromBuffer.GetName()], fromBuffer)
+	go publishMessages(ctx, startTime, messageValue, 300, 10, p, fromBuffer)
 
 	// wait until there is data in to buffer
 	for buffer.IsEmpty() {
@@ -369,6 +371,7 @@ func TestReduceDataForward_Sum(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -385,18 +388,18 @@ func TestReduceDataForward_Sum(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
-
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publishersMap := createPublisherForBuffer(ctx, toBuffer, pipelineName)
 	// create a fixed window of 2 minutes
 	window := fixed.NewFixed(2 * time.Minute)
 
 	var reduceDataForward *DataForward
-	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publishersMap,
 		window, WithReadBatchSize(10))
 	assert.NoError(t, err)
 
@@ -404,7 +407,7 @@ func TestReduceDataForward_Sum(t *testing.T) {
 	go reduceDataForward.Start()
 
 	// start the producer
-	publishMessages(ctx, startTime, messageValue, 300, 10, p[fromBuffer.GetName()], fromBuffer)
+	publishMessages(ctx, startTime, messageValue, 300, 10, p, fromBuffer)
 
 	// wait until there is data in to buffer
 	for buffer.IsEmpty() {
@@ -442,6 +445,7 @@ func TestReduceDataForward_Max(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -458,18 +462,19 @@ func TestReduceDataForward_Max(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publishersMap := createPublisherForBuffer(ctx, toBuffer, pipelineName)
 
 	// create a fixed window of 5 minutes
 	window := fixed.NewFixed(5 * time.Minute)
 
 	var reduceDataForward *DataForward
-	reduceDataForward, err = NewDataForward(ctx, MaxReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(ctx, MaxReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publishersMap,
 		window, WithReadBatchSize(10))
 	assert.NoError(t, err)
 
@@ -477,7 +482,7 @@ func TestReduceDataForward_Max(t *testing.T) {
 	go reduceDataForward.Start()
 
 	// start the producer
-	go publishMessages(ctx, startTime, messageValue, 600, 10, p[fromBuffer.GetName()], fromBuffer)
+	go publishMessages(ctx, startTime, messageValue, 600, 10, p, fromBuffer)
 
 	// wait until there is data in to buffer
 	for buffer.IsEmpty() {
@@ -515,6 +520,7 @@ func TestReduceDataForward_SumWithDifferentKeys(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -536,20 +542,20 @@ func TestReduceDataForward_SumWithDifferentKeys(t *testing.T) {
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
-
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publishersMap := createPublisherForBuffer(ctx, toBuffer, pipelineName)
 	// create a fixed window of 5 minutes
 	window := fixed.NewFixed(5 * time.Minute)
 
 	var reduceDataForward *DataForward
 
-	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publishersMap,
 		window, WithReadBatchSize(10))
 
 	assert.NoError(t, err)
 
 	// start the producer
-	go publishMessages(ctx, startTime, messages, 650, 10, p[fromBuffer.GetName()], fromBuffer)
+	go publishMessages(ctx, startTime, messages, 650, 10, p, fromBuffer)
 
 	// start the forwarder
 	go reduceDataForward.Start()
@@ -596,6 +602,7 @@ func TestReduceDataForward_NonKeyed(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -612,18 +619,19 @@ func TestReduceDataForward_NonKeyed(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publishersMap := createPublisherForBuffer(ctx, toBuffer, pipelineName)
 
 	// create a fixed window of 5 minutes
 	window := fixed.NewFixed(5 * time.Minute)
 
 	var reduceDataForward *DataForward
-	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, nonKeyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, nonKeyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publishersMap,
 		window, WithReadBatchSize(10))
 	assert.NoError(t, err)
 
@@ -631,7 +639,7 @@ func TestReduceDataForward_NonKeyed(t *testing.T) {
 	go reduceDataForward.Start()
 
 	// start the producer
-	go publishMessages(ctx, startTime, messages, 600, 10, p[fromBuffer.GetName()], fromBuffer)
+	go publishMessages(ctx, startTime, messages, 600, 10, p, fromBuffer)
 
 	// wait until there is data in to buffer
 	for buffer.IsEmpty() {
@@ -670,6 +678,7 @@ func TestDataForward_WithContextClose(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toBufferName   = "reduce-to-buffer"
+		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -697,20 +706,21 @@ func TestDataForward_WithContextClose(t *testing.T) {
 	assert.NoError(t, err)
 
 	// create in memory watermark publisher and fetcher
-	f, p := fetcherAndPublisher(cctx, toBuffer, fromBuffer, t.Name())
+	f, p := fetcherAndPublisher(cctx, fromBuffer, t.Name())
+	publishersMap := createPublisherForBuffer(cctx, toBuffer, pipelineName)
 
 	// create a fixed window of 5 minutes
 	window := fixed.NewFixed(5 * time.Minute)
 
 	var reduceDataForward *DataForward
-	reduceDataForward, err = NewDataForward(cctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+	reduceDataForward, err = NewDataForward(cctx, SumReduceTest{}, keyedVertex, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, publishersMap,
 		window, WithReadBatchSize(1))
 	assert.NoError(t, err)
 
 	// start the forwarder
 	go reduceDataForward.Start()
 	// window duration is 300s, we are sending only 200 messages with event time less than window end time, so the window will not be closed
-	publishMessages(cctx, startTime, messages, 200, 10, p[fromBuffer.GetName()], fromBuffer)
+	publishMessages(cctx, startTime, messages, 200, 10, p, fromBuffer)
 	// wait for the partitions to be created
 	for {
 		partitionsList := pbqManager.ListPartitions()
@@ -749,7 +759,7 @@ func TestDataForward_WithContextClose(t *testing.T) {
 }
 
 // fetcherAndPublisher creates watermark fetcher and publishers, and keeps the processors alive by sending heartbeats
-func fetcherAndPublisher(ctx context.Context, toBuffers map[string]isb.BufferWriter, fromBuffer *simplebuffer.InMemoryBuffer, key string) (fetch.Fetcher, map[string]publish.Publisher) {
+func fetcherAndPublisher(ctx context.Context, fromBuffer *simplebuffer.InMemoryBuffer, key string) (fetch.Fetcher, publish.Publisher) {
 
 	var (
 		keyspace     = key
@@ -762,20 +772,8 @@ func fetcherAndPublisher(ctx context.Context, toBuffers map[string]isb.BufferWri
 	hb, hbWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, hbBucketName)
 	ot, otWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, otBucketName)
 
-	publishers := make(map[string]publish.Publisher)
-
 	// publisher for source
 	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
-
-	// create publisher for to Buffers
-	for key := range toBuffers {
-		publishEntity := processor.NewProcessorEntity(key)
-		hb, _, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, key+"_PROCESSORS")
-		ot, _, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, key+"_OT")
-		p := publish.NewPublish(ctx, publishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
-		publishers[key] = p
-	}
-	publishers[fromBuffer.GetName()] = sourcePublisher
 
 	// publish heartbeat for the processors
 	go func() {
@@ -794,7 +792,21 @@ func fetcherAndPublisher(ctx context.Context, toBuffers map[string]isb.BufferWri
 	otWatcher, _ := inmem.NewInMemWatch(ctx, pipelineName, keyspace+"_OT", otWatcherCh)
 
 	var f = fetch.NewEdgeFetcher(ctx, fromBuffer.GetName(), wmstore.BuildWatermarkStoreWatcher(hbWatcher, otWatcher))
-	return f, publishers
+	return f, sourcePublisher
+}
+
+func createPublisherForBuffer(ctx context.Context, toBuffers map[string]isb.BufferWriter, pipelineName string) map[string]publish.Publisher {
+	publishers := make(map[string]publish.Publisher)
+
+	// create publisher for to Buffers
+	for key := range toBuffers {
+		publishEntity := processor.NewProcessorEntity(key)
+		hb, _, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, key+"_PROCESSORS")
+		ot, _, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, key+"_OT")
+		p := publish.NewPublish(ctx, publishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+		publishers[key] = p
+	}
+	return publishers
 }
 
 // buildMessagesForReduce builds test isb.Message which can be used for testing reduce.

--- a/pkg/reduce/pbq/pbq.go
+++ b/pkg/reduce/pbq/pbq.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"errors"
 	"strconv"
+	"sync"
 
 	"github.com/numaproj/numaflow/pkg/metrics"
 	"github.com/numaproj/numaflow/pkg/reduce/pbq/partition"
@@ -45,6 +46,7 @@ type PBQ struct {
 	options       *options
 	manager       *Manager
 	log           *zap.SugaredLogger
+	mu            sync.Mutex
 }
 
 var _ ReadWriteCloser = (*PBQ)(nil)
@@ -86,7 +88,13 @@ func (p *PBQ) CloseOfBook() {
 // Close is used by the writer to indicate close of context
 // we should flush pending messages to store
 func (p *PBQ) Close() error {
-	return p.store.Close()
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// we need a nil check because GC could have been invoked before close
+	if p.store != nil {
+		return p.store.Close()
+	}
+	return nil
 }
 
 // ReadCh exposes read channel to read messages from PBQ
@@ -98,6 +106,9 @@ func (p *PBQ) ReadCh() <-chan *isb.ReadMessage {
 // GC cleans up the PBQ and also the store associated with it. GC is invoked after the Reader (ProcessAndForward) has
 // finished forwarding the output to ISB.
 func (p *PBQ) GC() error {
+	// we need a lock because Close() and GC() can be invoked simultaneously
+	p.mu.Lock()
+	defer p.mu.Unlock()
 	p.store = nil
 	return p.manager.deregister(p.PartitionID)
 }

--- a/pkg/reduce/pbq/pbq.go
+++ b/pkg/reduce/pbq/pbq.go
@@ -90,7 +90,7 @@ func (p *PBQ) CloseOfBook() {
 func (p *PBQ) Close() error {
 	p.mu.Lock()
 	defer p.mu.Unlock()
-	// we need a nil check because GC could have been invoked before close
+	// we need a nil check because PBQ.GC could have been invoked before close
 	if p.store != nil {
 		return p.store.Close()
 	}
@@ -106,7 +106,8 @@ func (p *PBQ) ReadCh() <-chan *isb.ReadMessage {
 // GC cleans up the PBQ and also the store associated with it. GC is invoked after the Reader (ProcessAndForward) has
 // finished forwarding the output to ISB.
 func (p *PBQ) GC() error {
-	// we need a lock because Close() and GC() can be invoked simultaneously
+	// we need a lock because Close() and PBQ.GC() can be invoked simultaneously
+	// by shutdown routine(pbq.GC in case of ctx close) and pnf(pbq.Close after forwarding the result)
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	p.store = nil

